### DISCUSSION TITLE
0 DSPs

### DISCUSSION THREAD

**Discussion by vkvasan** [fastmachinelearning/hls4ml#698] [Q&A]
How do you force a neural network to take '0' DSPs in HLS4ML ? Also is there a way that we could implement BNN or TNN with HLS4ML APIs ?

**Comment by vloncar** [comment]
DSPs won't be used if the precision of the weights and activations is less than 10 bits. You can train BNNs and TNNs with [QKeras](https://github.com/google/qkeras)

**Reply by vkvasan** [reply]
I am having the precision less than 10 bits. The network is given below- 

 inputs = Input(shape=[98, ], name='qinput_features')
  x = QDense(20, kernel_quantizer=quantized_bits(6,0,alpha=1), bias_quantizer=quantized_bits(6,0,alpha=1), name='fc1')(inputs)
  x = QActivation(activation=quantized_relu(6), name='relu1')(x)
  x = QDense(1, kernel_quantizer=quantized_bits(6,0,alpha=1), bias_quantizer=quantized_bits(6,0,alpha=1), name='output')(x)
  x = Activation(activation="sigmoid", name='output_act')(x)
  model = Model(inputs, x)  

**If I change the input shape from 98 to 50 , the number of DSP slices increases from 0 to 33. Can you please explain why this happens ?**

**Reply by vloncar** [reply]
Look into the synthesis reports to see what exactly is using DSPs.
